internal/cmd/dev.go

```
@@ -89,16 +98,390 @@ func (s *statement) UnmarshalJSON(data []byte) error {
+			return nil, hranaErr(err)
+		}
+		defer rows.Close()
+
+		columns, _ := rows.Columns()
+		if columns == nil {
+			columns = []string{}
+		}
+
+		for _, col := range columns {
+			// Consistent with the response from SQLd
+			if strings.ToLower(col) == "null" {
+				col = "NULL"
+			}
+			c := hranaCol{col}
+			result.Cols = append(result.Cols, c)
+		}
+
+		for rows.Next() {
+			values := make([]interface{}, len(columns))
+			pointers := make([]interface{}, len(columns))
+			for i := range values {
+				pointers[i] = &values[i]
+			}
+			err := rows.Scan(pointers...)
+			if err != nil {
+				return nil, hranaErr(err)
+			}
+			row := make([]hranaValue, 0)
+			for _, val := range values {
+				switch v := val.(type) {
+				case nil:
+					row = append(row, hranaValue{Type: "null"})
+				case int64:
+					row = append(row, hranaValue{Type: "integer", Value: fmt.Sprintf("%d", v)})
+				case float64:
+					row = append(row, hranaValue{Type: "float", Value: v})
+				case []byte:
+					row = append(row, hranaValue{Type: "blob", Base64: string(v)})
+				case string:
+					row = append(row, hranaValue{Type: "text", Value: v})
+				default:
+					row = append(row, hranaValue{Type: "text", Value: v.(string)})
+				}
+			}
+			result.Rows = append(result.Rows, row)
+		}
 	}
+	return &result, nil
+}
+
+func executeStatement(c *gin.Context, db *sql.DB) {
+	var reqBody hranaExecuteReq
+	var result hranaSuccessResponse
+	err := c.BindJSON(&reqBody)
+	if err != nil {
+		c.JSON(http.StatusBadRequest, hranaErr(err))
+		return
+	}
+
+	res, herr := executeHranaStatement(reqBody.Stmt, db)
+	if herr != nil {
+		c.JSON(http.StatusBadRequest, herr)
+		return
+	}
+	result.Result = *res
+	c.JSON(http.StatusOK, result)
+}
+
+// 0 -> skipped
+// 1 -> error
+// 2 -> ok
+func evaluateCond(cond hranaCond, stmtRes []int) bool {
+	ty := strings.ToLower(cond.Type)
+	if ty == "ok" {
+		// malformed
+		if cond.Step == nil {
+			return false
+		}
+		step := *cond.Step
+		return stmtRes[step] == 2
```

Passing an out-of-range `step` will cause a panic. Other protocol violations (such as passing invalid `hranaValue`) will also cause a panic, so perhaps we can also panic when `cond.Step == nil`, rather than silently ignoring the error?